### PR TITLE
fix: job fails if Write node has no folders in filename

### PR DESCRIPTION
**Pull request by amznestebanpapp**
### What was the problem/requirement? (What/Why)

When submitting a job in Nuke that has a `Write` node with a filename that does not contain a folder, the job fails.
Example nk:
```
Root {
 inputs 0
 name colorbars.nk
 frame -3
 format "2048 1556 0 0 2048 1556 1 2K_Super_35(full-ap)"
 proxy_type scale
 proxy_format "1024 778 0 0 1024 778 1 1K_Super_35(full-ap)"
 colorManagement Nuke
 workingSpaceLUT linear
 monitorLut sRGB
 monitorOutLUT rec709
 int8Lut sRGB
 int16Lut sRGB
 logLut Cineon
 floatLut linear
}
ColorBars {
 inputs 0
 barintensity {{curve x1 0 x10 1}}
 name ColorBars1
 xpos -444
 ypos -375
}
set N6b5d800 [stack 0]
Viewer {
 frame -3
 frame_range 1-100
 monitorOutOutputTransform rec709
 name Viewer1
 xpos -444
 ypos -218
}
push $N6b5d800
Write {
 file output_####.exr
 file_type exr
 first_part rgba
 name Write1
 xpos -296
 ypos -259
}
```
Error:
```
2023/09/06 17:18:10-07:00 ADAPTOR_OUTPUT: STDERR: Traceback (most recent call last):
2023/09/06 17:18:10-07:00 ADAPTOR_OUTPUT: STDERR:   File "/tmp/ojio/session-c3d2e8510f7a4d1c879653acad8b2319rshv_oqi/session-env/packages/nuke/13.2/platform-linux/plugins/nuke_internal/callbacks.py", line 173, in beforeRender
2023/09/06 17:18:10-07:00 ADAPTOR_OUTPUT: STDERR:     _doCallbacks(beforeRenders)
2023/09/06 17:18:10-07:00 ADAPTOR_OUTPUT: STDERR:   File "/tmp/ojio/session-c3d2e8510f7a4d1c879653acad8b2319rshv_oqi/session-env/packages/nuke/13.2/platform-linux/plugins/nuke_internal/callbacks.py", line 46, in _doCallbacks
2023/09/06 17:18:10-07:00 ADAPTOR_OUTPUT: STDERR:     f[0](*f[1],**f[2])
2023/09/06 17:18:10-07:00 ADAPTOR_OUTPUT: STDERR:   File "/tmp/ojio/session-c3d2e8510f7a4d1c879653acad8b2319rshv_oqi/session-env/packages/deadline_cloud_for_nuke/0.9.0/_v/a/python/deadline/nuke_adaptor/NukeClient/nuke_client.py", line 48, in ensure_output_dir
2023/09/06 17:18:10-07:00 ADAPTOR_OUTPUT: STDERR:     os.makedirs(output_dir)
2023/09/06 17:18:10-07:00 ADAPTOR_OUTPUT: STDERR:   File "/tmp/ojio/session-c3d2e8510f7a4d1c879653acad8b2319rshv_oqi/session-env/packages/nuke/13.2/platform-linux/lib/python3.7/os.py", line 223, in makedirs
2023/09/06 17:18:10-07:00 ADAPTOR_OUTPUT: STDERR:     mkdir(name, mode)
2023/09/06 17:18:10-07:00 ADAPTOR_OUTPUT: STDERR: FileNotFoundError: [Errno 2] No such file or directory: ''
```

### What was the solution? (How)

Check that there is a folder in the filename before trying to create one.

### What is the impact of this change?

Despite uncommon in practice, a `Write` node does not necessarily has to contain a folder. A user could indicate an output filename without a folder. 

### How was this change tested?

Above nk script was used to submit jobs. A job with a filename that does not contain a folder was submitted successful with these changes.
Other changes were done to be able to upload wheels (also tested).

### Did you run the "Job Bundle Output Tests"? If not, why not? If so, paste the test results here.

The difference is because I am using a newer version of Nuke. No other differences found:
```
Timestamp: 2023-09-07T19:06:49.873239-07:00
Running job bundle output test: D:\ws\deadline-cloud-for-nuke\job_bundle_output_tests\multi-load-save

multi-load-save
Test failed, found differences
--- expected/parameter_values.yaml
+++ test/parameter_values.yaml
@@ -14,6 +14,6 @@
 - name: ProxyMode
   value: 'false'
 - name: NukeVersion
-  value: 13.2v4
+  value: 14.0v5
 - name: RezPackages
   value: nuke-13 deadline_cloud_for_nuke

Timestamp: 2023-09-07T19:18:28.632054-07:00
Running job bundle output test: D:\ws\deadline-cloud-for-nuke\job_bundle_output_tests\noise-saver

noise-saver
Test failed, found differences
--- expected/parameter_values.yaml
+++ test/parameter_values.yaml
@@ -14,6 +14,6 @@
 - name: ProxyMode
   value: 'false'
 - name: NukeVersion
-  value: 13.2v4
+  value: 14.0v5
 - name: RezPackages
   value: nuke-13 deadline_cloud_for_nuke

Failed 2 tests, succeeded 0.
Timestamp: 2023-09-07T19:19:00.837552-07:00
```

### Was this change documented?

No

### Is this a breaking change?

Yes, this is updating the dependencies to `deadline-cloud` and `openjobio`, its also changing the module name to `openjd` since that is what is being generated.

### CODE REVIEW

File: src/deadline/nuke_submitter/deadline_submitter_for_nuke.py
Comment (mwiebe) [inline src/deadline/nuke_submitter/deadline_submitter_for_nuke.py:216]: I think this needs logic when override_rez_packages is off as well. Then it would take the default value from the job template as the starting point.

File: src/deadline/nuke_adaptor/NukeClient/nuke_client.py
Comment (mwiebe) [inline src/deadline/nuke_adaptor/NukeClient/nuke_client.py:50]: If they do not contain folders, where is the file written? The code that constructs the FlatAssetReferences output directories list needs to know this for correct job attachments behavior. We should add this case to one of our job bundle output test nuke scripts.

File: pyproject.toml
Comment (mwiebe) [inline pyproject.toml]: We need to wait for stable closed-beta installers before we merge this dep update.

**Review by mwiebe — CHANGES_REQUESTED** [review]

File: DEVELOPMENT.md
Comment (mwiebe) [inline DEVELOPMENT.md]: Since you added the build_wheels.sh script, I think this should just call `./scripts/build_wheels.sh` now.

File: scripts/build_wheels.sh
Comment (mwiebe) [inline scripts/build_wheels.sh:3]: As a good habit, I like it when we add `set -euo pipefile` to the start of every bash script we write.

File: src/deadline/nuke_adaptor/NukeClient/nuke_client.py
Comment (agerveshi) [reply src/deadline/nuke_adaptor/NukeClient/nuke_client.py:50]: I believe the `project_directory` if set, otherwise the directory that Nuke was launched from.

Comment (amznestebanpapp) [reply src/deadline/nuke_adaptor/NukeClient/nuke_client.py:50]: This may be reason why I cannot download the results. Let me figure out how to get an absolute path here. Either using `project_directory` as @agerveshi is suggesting or using `cwd` if that does not exist. 
I will also test what it does locally in this scenario where `project_directory` is not set.There is also a "Script Directory" that creates an expression to set the project directory to the one where the script is.

File: src/deadline/nuke_submitter/deadline_submitter_for_nuke.py
Comment (amznestebanpapp) [reply src/deadline/nuke_submitter/deadline_submitter_for_nuke.py:216]: ok, I will pass a default value similar to what you did in the maya submitter and then do another PR to replicate your changes.

File: pyproject.toml
Comment (amznestebanpapp) [reply pyproject.toml]: ok, I will revert these changes for the time being

File: scripts/build_wheels.sh
Comment (amznestebanpapp) [reply scripts/build_wheels.sh:3]: ok, added it and also added it to `add_copyright_headers.sh` where I copied the header from.

File: src/deadline/nuke_adaptor/NukeClient/nuke_client.py
Comment (amznestebanpapp) [reply src/deadline/nuke_adaptor/NukeClient/nuke_client.py:50]: looks like `project_directory` is considered as part of `nuke.filename` and it even does the expansion when its the script directory:
```
2023/09/08 11:43:07-07:00 ADAPTOR_OUTPUT: STDOUT: output_filename = /tmp/ojio/session-967eca15f6de4c78a4a23285f4f05f7e5t4580s7/assetroot-ca41c57d5028a6a71bf7/output_####.exr
```
So, will just check if it is not abs, if it is not, will make it relative to `cwd`

File: scripts/build_wheels.sh
Comment (mwiebe) [reply scripts/build_wheels.sh:3]: Awesome

File: src/deadline/nuke_submitter/job_bundle_output_test_runner.py
Comment (mwiebe) [inline src/deadline/nuke_submitter/job_bundle_output_test_runner.py:231]: I suspect we should explicitly set the CWD to a known temporary directory different from the job bundle dir, to ensure this has predictable behavior.

**Review by mwiebe — APPROVED** [review]

File: src/deadline/nuke_submitter/job_bundle_output_test_runner.py
Comment (amznestebanpapp) [reply src/deadline/nuke_submitter/job_bundle_output_test_runner.py:231]: Not sure I am following the comment.

the CWD in my case was no the job bundle dir, but the path where I launch Nuke from (which happened to be the root of this repo). 
This currently works because `os.getcwd` does not change from the moment job attachments collects the paths to this point.

Can you clarify what is the unpredictable behavior you are concerned about?

Comment (mwiebe) [reply src/deadline/nuke_submitter/job_bundle_output_test_runner.py:231]: The CWD depends on how the user launched Nuke. It might contain an arbitrary set of files in it, that could be read or ignored by Nuke or scripts embedded in a Nuke scene. Creating a fresh temporary directory, and setting the CWD to it before running the tests eliminates that potential variability.

**Review by rmv — APPROVED** [review]
LGTM!

File: src/deadline/nuke_submitter/assets.py
Comment (jlangmann-aws) [inline src/deadline/nuke_submitter/assets.py:31]: Does it make sense to set this to the directory of the `nuke.scriptName()` instead of the directory from which Nuke was launched?

Comment (amznestebanpapp) [reply src/deadline/nuke_submitter/assets.py:31]: Nuke takes cwd as that path. Meaning, if we didnt define `project_directory` and have a relative file path in a `Write` node. The file will go to cwd. 
So here we have to take the same approach, otherwise you could have `project_directory` undefined and a relative path in a `Read` node and assume that you are relative to `nuke.scriptName`, but Nuke is realtive to cwd.
I have seen some cases where TDs set this up relative to the cwd and have artists launch the tool from a specific path (e.g. from within the "shot" path).

**Review by agerveshi — APPROVED** [review]

**Review by mwiebe — APPROVED** [review]

**Pull request merged**
